pam/internal/gdm/conversation_test.go

```diff
@@ -206,6 +206,128 @@ func TestSendData(t *testing.T) {
+		wantConvHandlerNotToBeCalled bool
+	}{
+		"Send valid data and return it back": {
+			inData: &Data{
+				Type:  DataType_hello,
+				Hello: &HelloData{Version: 12345},
+			},
+			outData:    &Data{Type: DataType_hello},
+			wantReturn: &Data{Type: DataType_hello},
+		},
+
+		// Error cases
+		"Error on invalid protocol": {
+			mayHitLeakSanitizer: true,
+			inBinReq: func() pam.BinaryConvRequester {
+				if pam_test.IsAddressSanitizerActive() {
+					return nil
+				}
+				invalidData := allocateJSONProtoMessage()
+				invalidData.init("testProto", 20, nil)
+				return pam.NewBinaryConvRequest(invalidData.encode(),
+					func(ptr pam.BinaryPointer) { (*jsonProtoMessage)(ptr).release() })
+			}(),
+			wantConvHandlerNotToBeCalled: true,
+			wantError:                    ErrProtoNotSupported,
+		},
+		"Error on unexpected JSON": {
+			mayHitLeakSanitizer: true,
+			inBinReq: func() pam.BinaryConvRequester {
+				if pam_test.IsAddressSanitizerActive() {
+					return nil
+				}
+				req, err := NewBinaryJSONProtoRequest([]byte("null"))
+				require.NoError(t, err)
+				return req
+			}(),
+			wantConvHandlerNotToBeCalled: true,
+			wantError:                    errors.New("syntax error"),
+		},
+		"Error on invalid Returned Data": {
+			inData: &Data{
+				Type:  DataType_hello,
+				Hello: &HelloData{Version: 12345},
+			},
+			outData:   &Data{},
+			wantError: pam.ErrConv,
+		},
+	}
+	for name, tc := range testCases {
+		tc := tc
+		name := name
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Cleanup(pam_test.MaybeDoLeakCheck)
+
+			if pam_test.IsAddressSanitizerActive() && tc.mayHitLeakSanitizer {
+				t.Skip("This test may cause false positive detection of leaks, so we ignore it")
+			}
+
+			convFuncCalled := false
+			var outData *Data
+			var outErr error
+			if tc.inBinReq != nil {
+				defer tc.inBinReq.Release()
+				outPtr, err := DataConversationFunc(func(d *Data) (*Data, error) {
+					convFuncCalled = true
+					requireEqualData(t, tc.inData, d)
+					if tc.outData != nil {
+						return tc.outData, nil
+					}
+					return nil, tc.wantError
+				}).RespondPAMBinary(tc.inBinReq.Pointer())
+
+				if err != nil {
+					require.Nil(t, outPtr)
+					outErr = err
+				} else {
+					json, err := decodeJSONProtoMessage(outPtr)
+					require.NoError(t, err)
+					defer (*jsonProtoMessage)(outPtr).release()
+					outData, err = NewDataFromJSON(json)
+					require.NoError(t, err)
+				}
+			} else {
+				mtx := pam_test.NewModuleTransactionDummy(DataConversationFunc(
+					func(data *Data) (*Data, error) {
+						convFuncCalled = true
+						requireEqualData(t, data, tc.inData)
+						if tc.outData != nil {
+							return tc.outData, nil
+						}
+						return nil, tc.wantError
+					}))
+				outData, outErr = SendData(mtx, tc.inData)
+			}
+			require.Equal(t, !convFuncCalled, tc.wantConvHandlerNotToBeCalled)
+
+			if tc.wantError != nil {
+				require.ErrorContains(t, outErr, tc.wantError.Error())
```

Comment:
Sadly we can't because we may have this JSON syntax error: `proto: syntax error (line 1:1): unexpected token null` and that's is emitted from JSON and it doesn't expose the error type, so I wasn't able to catch it otherwise.